### PR TITLE
feat: transform media min/max queries

### PR DESCRIPTION
**Pull request by agriffis**
## Summary

Wouldn't it be wonderful if xstyled magic applied to media queries? 🪄

This pull request replaces values in `min-width` and `max-width` media queries:

```js
const Foo = styled.div`
  @media (min-width: md) and (max-width: lg) {
    background-color: rebeccapurple;
  }
`
```

results in:

```css
@media (min-width 768px) and (max-width: 1023.98px) {
  background-color: rebeccapurple;
}
```

## Implementation notes

There are five commits here in deliberate sequence:

1. refactor
2. fixes for existing problems
3. feature
4. benchmark
5. docs

The refactoring uses the `trex` utility, which stands for tagged regular expressions, or tyrannosaurus rex if you prefer. It's a thin wrapper around `new RegExp()` to allow commenting. (`trex` is inspired by [XRegExp.tag](https://xregexp.com/api/#tag) but my own code—I might release it as a standalone lib eventually.)

The media query transform uses getters in a separate namespace: `mediaGetters`. This is because they would otherwise clash with existing prop getters by the same names, and they need to follow different rules. The other option would be to prefix them, for example `@media-min-width` but that seems kludgy and unnecessary.

One interesting thing is that `(min-width: lg)` transforms to `(min-width: 1024px)`, but the unitless `(min-width: 1024)` does not transform to `(min-width: 1024px)`. This might be a good thing, though, since if `rpxTransformers` were included then it would transform to rems, and that's not a great idea because [browsers always interpret media queries as 16px rems and ems regardless of styling](https://drafts.csswg.org/mediaqueries/#units).

## Test plan

I added new tests, both for existing code that wasn't directly tested (the `transform` function) and for the new feature.

## Benchmarking

Using the styled-components benchmark (`yarn bench styled-components`). The `benchme.bash` script runs it 20 times and drops the outliers before averaging.

<table><tr><th>baseline</th><th>unmodified benchmark</th><th>benchmark with media query</tr><tr><td>

```
✸ ./benchme.bash
0 39242 17087
1 40100 17920
2 38716 17170
3 40828 17661
4 40400 17549
5 38772 17379
6 40857 17532
7 39921 17908
8 40846 17969
9 40755 17274
10 40574 17766
11 40270 18355
12 40216 17797
13 39943 18011
14 40376 17694
15 39861 17890
16 40065 17197
17 40191 17905
18 39181 17052
19 39095 17274
styled-components avg = 40035
xstyled avg = 17610
```

</td><td>

```
✸ ./benchme.bash
0 40314 17251
1 40451 17481
2 39990 17376
3 40133 17941
4 40349 17603
5 40247 17631
6 40038 17814
7 39962 17594
8 39619 17304
9 39095 17617
10 40353 17676
11 40402 18128
12 39500 18008
13 39483 17513
14 40430 17548
15 40752 17592
16 40312 17641
17 39816 17983
18 40402 17475
19 40051 17697
styled-components avg = 40102
xstyled avg = 17638
```

</td><td>

```
✸ ./benchme.bash
0 38666 16033
1 39821 16506
2 39508 16166
3 40179 16048
4 40122 15939
5 38875 16036
6 40351 16448
7 39534 16120
8 40293 16428
9 39571 16421
10 41476 16373
11 39787 16120
12 40394 15862
13 39684 16194
14 39609 16016
15 40692 16122
16 40541 16002
17 39385 15983
18 39308 15719
19 39648 16072
styled-components avg = 39850
xstyled avg = 16132
```

</td></tr></table>

In summary:

- the new code doesn't change the speed of the unmodified benchmark (which doesn't include a media query)
- the modifed benchmark shows an expected small drop for both styled-components and xstyled

Side note: The benchmark mainly measures component creation, since that takes a lot more time relatively than element creation. In other words it's benchmarking the `transform` function more than anything else. But for most apps, component creation only runs at initialization, whereas element creation runs often during execution. If you move the component creation outside of the benchmarked function, xstyled and styled-components are effectively the same speed—naturally, since the output of `transform` is just a normal styled component.

### CODE REVIEW

**Comment by agriffis** [comment]
Ugh. I was confused by the master to main switch, and I still had a really old master/main branch locally. So I need to rebase this to current, and that's going to be a bit of work.

**Comment by agriffis** [comment]
Phewph, all rebased :sweat_smile:

**Comment by agriffis** [comment]
@gregberge I added one more commit to partially update the documentation.

It's "partial" because this section should change, too: https://xstyled.dev/docs/enhanced-styled-components/#responsive-utilities

It seems like the `breakpoints`, `up`, and `down` utilities might be deprecated by this new capability. I'm not sure why anybody would prefer those over directly writing the media query. Can you think of a reason? Otherwise the old responsive utilities seem to be on a similar path as the old `variant` utility over time.

**Comment by agriffis** [comment]
@gregberge Could you let me know when you anticipate reviewing this? Not rushing you, I just would like an idea of what to expect. Thanks!

File: benchmarks/styled-components.js [hunk 1]
```diff
@@ -6,32 +6,37 @@ const styled = require('styled-components')
...
 function runXstyled() {
   const Component = xstyled.default.div`
-    background: transparent;
-    border-radius: 3px;
-    border: 2px solid;
-    border-color: primary;
-    color: primary;
-    margin: 0 3;
-    padding: 1 3;
+    media (min-width: 1024px) {
```
Comment (gregberge) [inline benchmarks/styled-components.js:16]: @media ?

File: benchmarks/styled-components.js [hunk 2]
```diff
@@ -6,32 +6,37 @@ const styled = require('styled-components')
...
 function runStyled() {
   const Component = styled.default.div`
-    background: transparent;
-    border-radius: 3px;
-    border: 2px solid;
-    border-color: ${() => primary};
-    color: ${() => primary};
-    margin: 0 ${() => sp3};
-    padding: ${() => sp1} ${() => sp3};
+    media (min-width: ${() => md}) {
```
Comment (gregberge) [inline benchmarks/styled-components.js:31]: Same

File: packages/core/src/mediaGetters.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  getBreakpoints,
+  getBreakpointMin,
+  getBreakpointMax,
+} from '@xstyled/system'
+
+const getMediaWidth = (getBreakpointBound: Function) => (value: any) => (
+  props: object,
+) => {
+  const v = getBreakpointBound(getBreakpoints(props), value)
+  // getters return return null for smallest, undefined for not found, and
```
Comment (gregberge) [inline packages/core/src/mediaGetters.ts:11]: return

File: packages/core/src/mediaGetters.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  // otherwise a px string. Since we MUST emit a matcher at this point, use 0px
+  // for the smallest case (always true for min, never true for max).
+  return v === null ? '0px' : v || value
```
Comment (gregberge) [inline packages/core/src/mediaGetters.ts:14]: why not "0" ?

**Review by gregberge — CHANGES_REQUESTED** [review]
> The refactoring uses the `trex` utility, which stands for tagged regular expressions, or tyrannosaurus rex if you prefer. It's a thin wrapper around `new RegExp()` to allow commenting. (`trex` is inspired by [XRegExp.tag](https://xregexp.com/api/#tag) but my own code—I might release it as a standalone lib eventually.)

I really see the usage of `trex` in a project but in a library it is not a good idea: we need perf and small size. Feel free to add a comment with `trex` version commented but including it in the code just make it slower and heavier.

**Comment by gregberge** [comment]
Aside the small problem with `trex` it is a very good job! Thank you! Big feature :)

**Comment by agriffis** [comment]
Thanks @gregberge. I'll make those changes and force-push an update this afternoon.

**Comment by agriffis** [comment]
@gregberge I tried something to keep the inline comments. It seems worthwhile to me for understanding the intention of the regular expression components. Let me know what you think.

Since I force-pushed you can't easily see what changed, but here is a diff for you:

```diff
diff --git a/benchmarks/styled-components.js b/benchmarks/styled-components.js
index 08aed3e..1bd8127 100644
--- a/benchmarks/styled-components.js
+++ b/benchmarks/styled-components.js
@@ -13,7 +13,7 @@ const sp1 = '1px'
 
 function runXstyled() {
   const Component = xstyled.default.div`
-    media (min-width: 1024px) {
+    @media (min-width: 1024px) {
       background: transparent;
       border-radius: 3px;
       border: 2px solid;
@@ -28,7 +28,7 @@ function runXstyled() {
 
 function runStyled() {
   const Component = styled.default.div`
-    media (min-width: ${() => md}) {
+    @media (min-width: ${() => md}) {
       background: transparent;
       border-radius: 3px;
       border: 2px solid;
diff --git a/packages/core/src/mediaGetters.ts b/packages/core/src/mediaGetters.ts
index 1ed3a08..294ec11 100644
--- a/packages/core/src/mediaGetters.ts
+++ b/packages/core/src/mediaGetters.ts
@@ -8,10 +8,10 @@ const getMediaWidth = (getBreakpointBound: Function) => (value: any) => (
   props: object,
 ) => {
   const v = getBreakpointBound(getBreakpoints(props), value)
-  // getters return return null for smallest, undefined for not found, and
-  // otherwise a px string. Since we MUST emit a matcher at this point, use 0px
-  // for the smallest case (always true for min, never true for max).
-  return v === null ? '0px' : v || value
+  // getters return null for smallest, undefined for not found, and otherwise
+  // a px string. Since we MUST emit a matcher at this point, use 0px for the
+  // smallest case (always true for min, never true for max).
+  return v === null ? '0' : v || value
 }
 
 export const mediaGetters = {
diff --git a/packages/core/src/transform.test.ts b/packages/core/src/transform.test.ts
index b8d5311..fb64257 100644
--- a/packages/core/src/transform.test.ts
+++ b/packages/core/src/transform.test.ts
@@ -58,7 +58,7 @@ describe('#transform', () => {
 
   it('transforms smallest value media query', () => {
     expect(transformToStr('@media (min-width: sm) {}')).toBe(
-      '@media (min-width: 0px) {}',
+      '@media (min-width: 0) {}',
     )
   })
 
diff --git a/packages/core/src/transform.ts b/packages/core/src/transform.ts
index e565417..7739398 100644
--- a/packages/core/src/transform.ts
+++ b/packages/core/src/transform.ts
@@ -1,36 +1,35 @@
 /* eslint-disable no-continue, no-loop-func, no-cond-assign */
-import { trex } from '@xstyled/util'
 import { mediaGetters } from './mediaGetters'
 import { propGetters } from './propGetters'
 
 // prop name is an ident: word chars, underscore and dash.
-const PROP_CHAR = trex('s')`[-\w]`
+const PROP_CHAR = `[-\\w]`
 
 // prop value consists of non-semis unless backslash-escaped.
-const VALUE_CHAR = trex('s')`(?:\\.|[^\\;])`
+const VALUE_CHAR = `(?:\\\\.|[^\\\\;])`
 
-const PROP_REGEXP = trex('gsx')`
-  (${PROP_CHAR}+)    # capture prop name
-  (\s*:\s*)          # colon & whitespace
-  (?=\S)             # prop value starts with non-whitespace
-  (${VALUE_CHAR}*?)  # capture prop value (non-greedy)
-  (\s*!important)?   # capture !important
-  (\s*;)             # semi & whitespace
-`
+// prettier-ignore
+const PROP_PATT = (
+  `(${PROP_CHAR}+)` +   // capture prop name
+  `(\\s*:\\s*)` +       // colon & whitespace
+  `(?=\\S)` +           // prop value starts with non-whitespace
+  `(${VALUE_CHAR}*?)` + // capture prop value (non-greedy)
+  `(\\s*!important)?` + // capture !important
+  `(\\s*;)`             // semi & whitespace
+)
 
 // simplistic but workable media query value.
-const MEDIA_CHAR = trex('s')`[^{]`
+const MEDIA_CHAR = `[^{]`
 
-const MEDIA_REGEXP = trex('gsx')`
-  (@media\b\s*)      # start of media query
-  (?=\S)             # value starts with non-whitespace
-  (${MEDIA_CHAR}+?)  # capture queries (non-greedy)
-  (\s*\{)            # brace & whitespace
-`
+// prettier-ignore
+const MEDIA_PATT = (
+  `(@media\\b\\s*)` +   // start of media query
+  `(?=\\S)` +           // value starts with non-whitespace
+  `(${MEDIA_CHAR}+?)` + // capture queries (non-greedy)
+  `(\\s*\\{)`           // brace & whitespace
+)
 
-const MATCH_REGEXP = trex('gsx')`
-  (?:${PROP_REGEXP}|${MEDIA_REGEXP})
-`
+const MATCH_REGEXP = new RegExp(`(?:${PROP_PATT}|${MEDIA_PATT})`, `gs`)
 
 export function transform(rawValue: any) {
   if (typeof rawValue !== 'string') return rawValue
@@ -62,13 +61,15 @@ export function transform(rawValue: any) {
 }
 
 // media query prop/value pairs such as (min-width: 1024px)
-const QUERY_REGEXP = trex('gsx')`
-  (\(\s*)            # open paren, whitespace
-  (${PROP_CHAR}+)    # capture prop name
-  (\s*:\s*)          # colon & whitespace
-  ([^\)]*?)          # capture prop value (non-greedy)
-  (\s*\))            # close paren, whitespace
-`
+// prettier-ignore
+const QUERY_REGEXP = new RegExp(
+  `(\\(\\s*)` +       // open paren, whitespace
+  `(${PROP_CHAR}+)` + // capture prop name
+  `(\\s*:\\s*)` +     // colon & whitespace
+  `([^\\)]*?)` +      // capture prop value (non-greedy)
+  `(\\s*\\))`,        // close paren, whitespace
+  `gs`
+)
 
 function mediaTransform(rawValue: string) {
   let matches
diff --git a/packages/util/src/index.test.ts b/packages/util/src/index.test.ts
index 6c685a6..58e5722 100644
--- a/packages/util/src/index.test.ts
+++ b/packages/util/src/index.test.ts
@@ -12,7 +12,6 @@ import {
   cascade,
   getThemeValue,
   omit,
-  trex,
 } from '.'
 
 describe('util', () => {
@@ -163,38 +162,4 @@ describe('util', () => {
       expect(omit({ a: 'x', b: 'y' }, ['b'])).toEqual({ a: 'x' })
     })
   })
-
-  describe('#trex', () => {
-    it('makes a regexp', () => {
-      const r = trex`foo\sbar`
-      expect(r).toBeInstanceOf(RegExp)
-      expect(r.source).toEqual(/foo\sbar/.source)
-      expect(r.flags).toEqual('')
-    })
-
-    it('handles flags', () => {
-      const r = trex('gs')``
-      expect(r.flags).toEqual('gs')
-    })
-
-    it('handles the x flag', () => {
-      const r = trex('gx')`
-        ^ ab  # this is a comment
-        c d$
-        `
-      expect(r.source).toEqual('^abcd$')
-      expect(r.flags).toEqual('g')
-    })
-
-    it("doesn't need extra backslashes", () => {
-      const r1 = trex`foo\sbar`
-      const r2 = new RegExp('foo\\sbar')
-      expect(r1.source).toEqual(r2.source)
-    })
-
-    it('interpolates strings and regexes', () => {
-      const r = trex`${'foo\\sfoo'}${/bar\sbaz/}`
-      expect(r.source).toEqual('foo\\sfoobar\\sbaz')
-    })
-  })
 })
diff --git a/packages/util/src/index.ts b/packages/util/src/index.ts
index 033fb90..24f18da 100644
--- a/packages/util/src/index.ts
+++ b/packages/util/src/index.ts
@@ -208,45 +208,3 @@ function flattenDown(array: any[], result: any[]) {
 export function flatten(array: any[]) {
   return flattenDown(array, [])
 }
-
-function assertCompatibleFlags(f1: string, f2: string) {
-  ;['i', 'm', 's', 'u'].forEach((flag) => {
-    if (f1.includes(flag) !== f2.includes(flag)) {
-      throw new Error(`incompatible regex flags: ${f1} vs ${f2}`)
-    }
-  })
-}
-
-/**
- * Make a RegExp from a tagged template literal, with support for the x flag
- * (whitespace and comments) and interpolation of other RegExps.
- * @param flags and/or template literal
- */
-export function trex(flags: string = '', ...args: Array<any>): Function {
-  if (Array.isArray(flags)) {
-    // trex called without flags, i.e. trex`pattern`
-    return trex()(flags, ...args)
-  }
-  // @ts-ignore
-  return (strings, ...values): RegExp => {
-    let patt = strings.raw[0]
-    values.forEach((v, i) => {
-      if (v === null || v === undefined) {
-        v = ''
-      } else if (v.source !== undefined) {
-        // interpolate another RegExp
-        assertCompatibleFlags(flags, v.flags)
-        v = v.source
-      }
-      patt += v + strings.raw[i + 1]
-    })
-    if (flags.includes('x')) {
-      // remove comments
-      patt = patt.replace(/\s#.*/g, '')
-      // remove whitespace except backslash-escaped
-      patt = patt.replace(/(?:(\\.)|\s)/gs, '$1')
-      flags = flags.replace(/x/g, '')
-    }
-    return new RegExp(patt, flags)
-  }
-}
```

**Review by gregberge — APPROVED** [review]

**Comment by gregberge** [comment]
Thanks, it is good for me!
